src/wp-admin/includes/update-core.php

```diff
@@ -1708,3 +1750,123 @@ function _upgrade_590_force_deactivate_incompatible_plugins() {
 		deactivate_plugins( array( 'gutenberg/gutenberg.php' ), true );
 	}
 }
+
+/**
+ * The upgrader uses the classes in the outgoing version of WordPress.
+ *
+ * If a class is renamed in the incoming version, the upgrader will no longer
+ * be able to find the outgoing class when it is needed.
+ *
+ * This will cause a fatal error. To prevent this, we can preload the
+ * outgoing classes that we need for the upgrade into memory.
+ *
+ * This MUST be done at the beginning of `update_core()`.
+ *
+ * @since 6.0.0
+ *
+ * @global WP_Filesystem_Base $wp_filesystem  WordPress filesystem subclass.
+ * @global string             $wp_version     The current WordPress version.
+ *
+ * @param string $wp_base                     Base path of the WordPress installation.
+ * @param array  $preload_paths               An array of paths to preload, relative to $wp_base.
+ * @param array  $skip_paths                  An array of paths to skip when preloading, relative to $wp_base.
+ * @return true|WP_Error True on success, or WP_Error on failure.
+ */
+function preload_outgoing_version_files( $wp_base, array $preload_paths, $skip_paths = array() ) {
+	global $wp_filesystem, $wp_version;
+
+	if ( ! $wp_filesystem || ! is_object( $wp_filesystem ) ) {
+		return new WP_Error( 'fs_unavailable', __( 'Could not access filesystem.' ) );
+	}
+
+	if ( empty( $preload_paths ) ) {
+		return new WP_Error( 'empty_paths', __( 'The preload paths list is empty.' ) );
+	}
+
+	$errors = new WP_Error();
+
+	// Preload files.
+	$dangerous_paths = array( '', '.', '/' );
+
+	foreach ( $preload_paths as $path ) {
```

Comment:
take a look at `validate_file()`, I suspect you are doubling up a few things here.

Reply:
Nice @peterwilsoncc, hadn't thought of that at the time. `validate_file()` should work nicely here.